### PR TITLE
feat(speed): make the dtt faster for many small files

### PR DESCRIPTION
**Pull request by ProfOak**
Used the change in the API to download uncompressed tarfiles (API-175)
done by grouping the smaller files together in multiple larger, but
less, groups of downloads.

This change depends on:
https://github.com/NCI-GDC/gdcapi/pull/666
https://github.com/LabAdvComp/parcel/pull/93

@NCI-GDC/ucdevs r?

### CODE REVIEW

File: gdc_client/download/parser.py [hunk 1]
```diff
@@ -66,62 +69,97 @@ def get_client(args, token, **_kwargs):
...
         if should_retry.lower() == 'y':
             log.info('{} retries remaning...'.format(retry_amount))
-            log.info('Retrying download... {} in {} seconds'.format(uuid, wait_time))
+            log.info('Retrying download... {} in {} seconds'.format(url, wait_time))
             retry_amount -= 1
             time.sleep(wait_time)
             # client.download_files accepts a list of uuids to download
             # but we want to only try one at a time
-            _, e = client.download_files([uuid])
+            _, e = client.download_files([url])
```
Comment (millerjs) [inline gdc_client/download/parser.py]: nitpick: `e` is a error? could be a more descriptive variable name

File: gdc_client/query/index.py [hunk 2]
```diff
@@ -29,16 +34,115 @@ def get_annotations(self, file_id):
...
+                    if af:
+                        extra_files.append(af)
+                except:
+                    log.warn('Unable to find annotation files for {0}'.format(uuid))
+
+        # now the list of UUIDs contain the related files and annotations
+        # and can be grouped into bulk tarfile downloads if applicable
+        ids |= extra_files # set union
+
+        AVAILABLE_SIZES = {
```
Comment (millerjs) [inline gdc_client/query/index.py]: nitpick: nice to have options, but this is deadcode

File: gdc_client/query/index.py [hunk 3]
```diff
@@ -29,16 +34,115 @@ def get_annotations(self, file_id):
...
+                'medium': 134217728, # 128MB
+                'large': 268435456,  # 256MB
+        }
+        MAX_BUNDLE_LEN = AVAILABLE_SIZES['xsmall']
+
+	filesize_query = {
+            'fields': 'file_id,file_size',
+            'filters': '{"op":"and","content":[{"op":"in","content":{"field":"files.file_id","value":["' + '","'.join(ids) + '"]}}]}',
+            'from': '1',
+            'size':str(len(ids)),       # one big ol' request
```
Comment (millerjs) [inline gdc_client/query/index.py]: hm, curious why this `size` is a string here

File: gdc_client/query/index.py [hunk 4]
```diff
@@ -29,16 +34,115 @@ def get_annotations(self, file_id):
...
+                'xsmall': 33554432,  #  32MB
+                'small': 67108864,   #  64MB
+                'medium': 134217728, # 128MB
+                'large': 268435456,  # 256MB
+        }
+        MAX_BUNDLE_LEN = AVAILABLE_SIZES['xsmall']
+
+	filesize_query = {
+            'fields': 'file_id,file_size',
+            'filters': '{"op":"and","content":[{"op":"in","content":{"field":"files.file_id","value":["' + '","'.join(ids) + '"]}}]}',
```
Comment (millerjs) [inline gdc_client/query/index.py]: nitpick: long line could be broken up

File: gdc_client/download/client.py [hunk 5]
```diff
@@ -14,112 +15,87 @@
...
+                # {'ids': ['id1', 'id2'..., 'idn']}
+                ids = {"ids": s}
+
+                # using a POST request lets us avoid the MAX URL character length limit
+                r = requests.post(tarfile_url, stream=True, verify=self.verify, json=ids)
+                if r.status_code == requests.codes.ok:
+
+                    # {'content-disposition': 'filename=the_actual_filename.tar'}j
+                    filename = r.headers.get('content-disposition')
+                    filename = os.path.join(self.directory, filename.split('=')[1])
```
Comment (ProfOak) [inline gdc_client/download/client.py]: If the response code is a 200, I don't think there is a possibility that there would be no filename. It's probably better to handle it. Too bad there's no :defense: emoji.

**Pull request merged**